tools/data-api/internal/repositories/services.go

```diff
@@ -353,59 +164,244 @@ func (s *ServicesRepositoryImpl) ProcessResourceDefinitions(serviceName string,
 	resourceDefinition := ServiceApiVersionResourceDetails{}
 	resourceSchema := ResourceSchema{}
 	constants := make(map[string]ConstantDetails, 0)
+	models := make(map[string]ModelDetails)
+	operations := make(map[string]ResourceOperations)
+	resourceIds := make(map[string]ResourceIdDefinition)
 
-	// TODO rename this var so we can import path to use path.Join
-	path := fmt.Sprintf("%s/%s/%s/%s", s.directory, serviceName, version, resource)
-	files, err := os.ReadDir(path)
+	resourcePath := path.Join(s.directory, serviceName, version, resource)
+	files, err := os.ReadDir(resourcePath)
 	if err != nil {
-		return nil, fmt.Errorf("retrieving definitions under %s: %+v", path, err)
+		return nil, fmt.Errorf("retrieving definitions under %s: %+v", resourcePath, err)
 	}
 
 	for _, file := range files {
 		if file.IsDir() {
 			continue
 		}
 
+		definitionPath := path.Join(resourcePath, file.Name())
 		definitionType, definitionName, err := getDefinitionInfo(file.Name())
 		if err != nil {
 			return nil, err
 		}
-		// we lower case these so that it's compatible with other OS e.g. Windows
+
+		// we lower case this comparison so that it's compatible with other OS e.g. Windows
 		switch strings.ToLower(definitionType) {
+		// Ordering here is important, all Constants need to be processed first, then all Models
```

Comment:
Could you elaborate on the potential downstream issues that could occur if we didn't duplicate the validation from saving the API definitions? I currently can't wrap my head around what we would be guarding against (except for issues in handwritten services). I feel like duplication of the validation is unnecessary bloating in the mapping for the data API code.

Furthermore I still don't understand why we have an intermediate mapping stage in the data API, it currently goes:

(for reference -> represents a mapping from a struct to a struct in the crude "diagrams" below)

**_data api models (our JSON files) -> ServiceDetail struct in the Data API -> v1 endpoints_**

The reasoning given the last time this was discussed is to maintain backward compatibility and to allow the exposure of different versioned endpoints (e.g. v2 endpoints?).

But I as far as I can tell we have the freedom to map the data api models into as many endpoints as we wish so omitting the intermediate `ServiceDetails` would achieve the same thing?

**_data api models (our JSON files) -> v1 endpoints_**

Reply:
> Furthermore I still don't understand why we have an intermediate mapping stage in the data API

We recently discussed outputting the Tests into `*.hcl` files - which (already) means the data represented on disk isn't going to be 1:1 with the models we're wanting to return. For example, it might seem like a safe assumption that we can rely on the existence of tests in the structure `tests/{name}.hcl` - however the test setup also supports custom tests with multiple stages (see below) - as such the files persisted on disk are going to want to contain the path to the HCL file for that test (stage):

```json
{
  "tests": {
    "basic": "tests/basic.hcl",
    "complete": "tests/complete.hcl",
    "requiresImport": "tests/requiresImport.hcl",
    "other": {
      "first": [
        "tests/first-01.hcl",
        "tests/first-02.hcl"
      ]
    }
  }
}
```

However when returning the `TerraformResourceTestsDefinition` struct from the repository we shouldn't expose the file names (since these are only an implementation detail) - so mapping this data allows us to instead return the test configurations instead (so that other parts of the codebase don't rely on the file path) - giving a clear separation of concerns here.

Note that whilst we aren't using the custom test infrastructure today, it's hooked up through the codebase with the intention to be used for more complex resources, so we should continue to support that so we can utilize it as needed.

> The reasoning given the last time this was discussed is to maintain backward compatibility and to allow the exposure of different versioned endpoints (e.g. v2 endpoints?).

In addition to above - having the `repositories` package define the model which gets returned means that things are considerably more flexible, since we can both:

1) Change the backing store (e.g. to use a database [like we discussed previously], using a different set of data) - without rewriting the rest of the codebase.

2) Change the shape of the data in a more gradual fashion (e.g. splitting `Reference` into `ConstantReference` and `ModelReference` - and gradually updating usages to account for that)

We are still going to need to map from the types used in the `repository` package onto the `v1` endpoints, since as you've mentioned these are versioned endpoints that other tooling will be relying on - so doing this makes it easier to provide compatibility guarantees there.

Which I guess follows into:

> Could you elaborate on the potential downstream issues that could occur if we didn't duplicate the validation from saving the API definitions? 

At this point in time the Importer(s) are writing the API Definitions to disk (to keep this the same as the V1 Data API) - however we've talked about having the Data API be the only tool which manages the files in the future - meaning that it's likely we could update this to use a POST endpoint in the future (for example).

As such having this validation this validation:

1. Continues providing compatibility guarantees at each stage of Pandora's pipeline (since the downstream consumers can rely on the shape/payload) - which makes merging far less risky and means we have faster feedback (since if there's an error in the pipeline we can fix it and move forward). This also means that data issues can be caught in the part of the project where the issue exists - so (for example) the Generators can rely on the data from the Data API.
2. By being located in the Repository, which is loaded when the API is launched - and not relegated to a unit test, this validation becomes a mandatory code path - meaning that data issues are far more likely to be caught.
3. This also allows the `watch` command to provide faster feedback too - similar to 1 and 2 above.
4. Allows the Data API to lean on this when inserting/updating/deleting data in the future, when the Importer POSTs data to the Data API.

It's worth noting that Pandora initially didn't have these checks, which meant that diagnosing problems took far longer, which is why these got added to V1 (and in the other places) since we were losing time to it - from memory this has previously caught:

* Object Definitions being incorrect
* The Data API not correctly serving Models which were used as Discriminated Types, meaning these were output as regular models by mistake.
* The Data API not correctly serving Constants used in the (Request) Options object.

As such whilst this does increase the complexity here (since we're going to need to maintain the list of Constants persisted to disk and transform those here too), doing so allows us to lean on these errors rather than requiring us to jump into the debugger to track down the source of data issues, providing for far faster feedback (which also gets run in a unit test at PR time, when new data is generated).

Finally, longer term the Data API is likely to become the sole maintainer of this data, meaning that we'd likely want the same logic to validate the data that gets persisted to disk (handling generated data), in addition to when the data is loaded from disk (handling both generated and handwritten data) - meaning that this will become centralised - but that's one for the future.